File: intime/app/src/main/java/com/vpe_soft/intime/intime/recyclerview/TaskRecyclerViewAdapter.java
package com.vpe_soft.intime.intime.recyclerview;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.cardview.widget.CardView;

import com.vpe_soft.intime.intime.R;
import com.vpe_soft.intime.intime.activity.Dimensions;
import com.vpe_soft.intime.intime.database.DatabaseUtil;
import com.vpe_soft.intime.intime.receiver.AlarmUtil;
import com.vpe_soft.intime.intime.view.ViewUtil;

import java.util.Locale;

public class TaskRecyclerViewAdapter extends RecyclerViewCursorAdapter<TaskRecyclerViewAdapter.TaskRecyclerViewVH>{

    public TaskRecyclerViewAdapter(Context context, Cursor cursor, Locale locale) {
        super(context, locale);
        setupCursorAdapter(cursor, 0, R.layout.task_item, false);
    }

    @Override
    public TaskRecyclerViewVH onCreateViewHolder(ViewGroup viewGroup, int i) {
        View itemView = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.task_item,viewGroup,false);
        return new TaskRecyclerViewVH(itemView);
    }

    @Override
    public void onBindViewHolder(TaskRecyclerViewVH holder, int position) {
        cursorAdapter.getCursor().moveToPosition(position);
        setViewHolder(holder);
        cursorAdapter.bindView(null, context, cursorAdapter.getCursor());
    }

    public class TaskRecyclerViewVH extends RecyclerViewCursorViewHolder {

        public TextView title;
        public TextView date;
        public CardView card;
        public LinearLayout indicator;
        public LinearLayout help;
        private final String TAG = "TaskRecyclerViewVH";

        public TaskRecyclerViewVH(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.title);
            date = itemView.findViewById(R.id.description);
            card = itemView.findViewById(R.id.linear1);
            indicator = itemView.findViewById(R.id.indicator);
            help = itemView.findViewById(R.id.help);
        }

        @Override
        public void bindCursor(Cursor cursor) {
            Log.d(TAG, "bindCursor");
            String description = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseUtil.DESCRIPTION_FIELD));
            long nextAlarm = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseUtil.NEXT_ALARM_FIELD));
            long nextCaution = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseUtil.NEXT_CAUTION_FIELD));
            final long id = cursor.getLong(cursor.getColumnIndexOrThrow("_id"));
            final int pos = cursor.getPosition();
            // 0 - not ready (white), 1 - almost (yellow), 2 - ready (red)
            int phase = System.currentTimeMillis() > nextCaution ? System.currentTimeMillis() > nextAlarm ? 2 : 1 : 0;
            card.setRadius(0);
            card.setCardBackgroundColor(Color.WHITE);
            card.setCardElevation(0);
            if (!mainActivity.isDefaultViewOutlineProviderSet) {
                mainActivity.isDefaultViewOutlineProviderSet = true;
                mainActivity.setDefaultViewOutlineProvider(card.getOutlineProvider());
            }
            card.setOutlineProvider(null);
            GradientDrawable gradientDrawable1 = new GradientDrawable();
            gradientDrawable1.setCornerRadius(Dimensions.indicatorCornerRadius);
            GradientDrawable gradientDrawable2 = new GradientDrawable();
            gradientDrawable2.setCornerRadius(Dimensions.indicatorCornerRadius);
            gradientDrawable2.setColor(Color.WHITE);
            switch(phase){
                case 0:
                    gradientDrawable1.setColor(colors.cardIndicatorNeutral);
                    break;
                case 1:
                    gradientDrawable1.setColor(colors.cardIndicatorAlmost);
                    break;
                case 2:
                    gradientDrawable1.setColor(colors.cardIndicatorReady);
                    break;
            }
            indicator.setBackground(gradientDrawable1);
            help.setBackground(gradientDrawable2);
            title.setText(description);
            date.setText(AlarmUtil.getDateFromNextAlarm(locale, nextAlarm));
                 /*+ "\n"
                    + "next_alarm " + nextAlarm + "\n"
                    + "next_caution " + nextCaution + "\n"
                    + "id " + id + "\n"
                    + "pos " + pos + "\n"
                    + "phase " + phase + "\n");*/
            title.setTypeface(ViewUtil.getTypeface(context), Typeface.NORMAL);
            date.setTypeface(ViewUtil.getTypeface(context), Typeface.NORMAL);
            card.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View view) {
                    mainActivity.onItemLongClicked(id, pos);
                    return true;
                }
            });
        }
    }
}

File: intime/app/src/main/java/com/vpe_soft/intime/intime/recyclerview/TaskRecyclerViewAdapter.kt
@file:Suppress("MemberVisibilityCanBePrivate")

package com.vpe_soft.intime.intime.recyclerview

import android.content.Context
import android.database.Cursor
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.cardview.widget.CardView
import com.vpe_soft.intime.intime.R
import com.vpe_soft.intime.intime.activity.cardIndicatorAlmost
import com.vpe_soft.intime.intime.activity.cardIndicatorNeutral
import com.vpe_soft.intime.intime.activity.cardIndicatorReady
import com.vpe_soft.intime.intime.activity.indicatorCornerRadius
import com.vpe_soft.intime.intime.kotlin.*
import com.vpe_soft.intime.intime.receiver.getDateFromNextAlarm

class TaskRecyclerViewAdapter(context: Context) :
    RecyclerViewCursorAdapter<TaskRecyclerViewAdapter.TaskRecyclerViewVH>(context), Taggable {

    override val tag = "TaskRecyclerViewAdapter"

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): TaskRecyclerViewVH =
        TaskRecyclerViewVH(
            LayoutInflater.from(context).inflate(R.layout.task_item, viewGroup, false)
        )


    override fun onBindViewHolder(holder: TaskRecyclerViewVH, position: Int) {
        cursorAdapter.cursor.moveToPosition(position)
        setViewHolder(holder)
        cursorAdapter.bindView(null, context, cursorAdapter.cursor)
    }

    inner class TaskRecyclerViewVH(itemView: View) : RecyclerViewCursorViewHolder(itemView), Taggable {
        val title: TextView = itemView.findViewById(R.id.title)
        val date: TextView = itemView.findViewById(R.id.description)
        val card: CardView = itemView.findViewById(R.id.card)
        val indicator: LinearLayout = itemView.findViewById(R.id.indicator)
        val help: LinearLayout = itemView.findViewById(R.id.help)
        override val tag = "TaskRecyclerViewVH"

        override fun bindCursor(cursor: Cursor) {
            log("bindCursor")
            val description = cursor.getString(cursor.getColumnIndexOrThrow("description"))
            val nextAlarm = cursor.getLong(cursor.getColumnIndexOrThrow("next_alarm"))
            val nextCaution = cursor.getLong(cursor.getColumnIndexOrThrow("next_caution"))
            val id = cursor.getLong(cursor.getColumnIndexOrThrow("_id"))
            val pos = cursor.position
            // 0 - not ready (white), 1 - almost (yellow), 2 - ready (red)
            val millis = System.currentTimeMillis()
            val phase = if (millis > nextCaution) if (millis > nextAlarm) 2 else 1 else 0
            with(card) {
                radius = 0f
                setCardBackgroundColor(Color.WHITE)
                cardElevation = 0f
                outlineProvider = null
            }
            with(mainActivity) {
                if (!isDefaultViewOutlineProviderSet) {
                    isDefaultViewOutlineProviderSet = true
                    setDefaultViewOutlineProvider(card.outlineProvider)
                }
            }
            val gradientDrawable1 = GradientDrawable()
            gradientDrawable1.cornerRadius = indicatorCornerRadius
            val gradientDrawable2 = GradientDrawable()
            with(gradientDrawable2) {
                cornerRadius = indicatorCornerRadius
                setColor(Color.WHITE)
            }
            when (phase) {
                0 -> gradientDrawable1.setColor(context.cardIndicatorNeutral)
                1 -> gradientDrawable1.setColor(context.cardIndicatorAlmost)
                2 -> gradientDrawable1.setColor(context.cardIndicatorReady)
            }
            indicator.background = gradientDrawable1
            help.background = gradientDrawable2
            title.text = description
            date.text = getDateFromNextAlarm(context.locale, nextAlarm)
            card.longClickListener = {
                mainActivity.onItemLongClicked(id, pos)
            }
        }
    }

    init {
        setupCursorAdapter(context.cursor(), R.layout.task_item)
    }
}
